File: python/src/threads/evals/simulate_prefix.py
"""The prefix of a simulated case (spec lane 32, B.1): the real turns before the saved one.
Continue them when the agent's config_hash is unchanged -- the case log is imported and the
conversation goes on, so the prefix costs no model call -- and re-drive their inputs when it
changed, so a changed agent is graded on the whole path."""

from dataclasses import dataclass
from typing import Literal

from threads.agents.definition import dry_pin
from threads.agents.results import Thread
from threads.agents.store import open_store
from threads.evals.case_dir import CaseDir
from threads.evals.checks import CaseLog
from threads.evals.ledger import cost_bound
from threads.evals.live_env import Env, EvalAgent
from threads.log import ThreadStartedEvent, UserInputEvent
from threads.result import Ok
from threads.store import verify_export


@dataclass(frozen=True, slots=True)
class Plan:
    mode: Literal["continued", "redriven", "none"]
    redrive: tuple[str, ...] = ()
    """Recorded inputs to re-send before the opener; empty unless the prefix is re-driven."""
    turns: int = 0
    thread: Thread | None = None
    """The imported thread to continue; None when the agent starts a new one."""
    since: int = 0
    """Events at or below this seq are the imported prefix, not this conversation's."""
    base_cost_nanos: int = 0


def prefix_texts(log: CaseLog) -> tuple[str | None, ...]:
    """The recorded user inputs before the saved turn, in log order. A content-part
    input has no text, and a live run skips the case."""
    return tuple(
        e.data.text if isinstance(e.data.text, str) else None
        for e in log.events
        if isinstance(e, UserInputEvent)
    )


def _recorded_hash(log: CaseLog) -> str | None:
    started = next((e for e in log.events if isinstance(e, ThreadStartedEvent)), None)
    return None if started is None else started.data.config_hash


async def _import_case(case: CaseDir, env: Env) -> Plan | None:
    """Imports the case log into the eval store, so the conversation continues the real thread."""
    verified = verify_export(case.log, case.meta.clock.now)
    if not isinstance(verified, Ok):
        return None
    sq = await open_store(env.store)
    for artifact in case.artifacts:
        await sq.put_artifact(artifact)
    imported = await sq.import_log(verified.value)
    if not isinstance(imported, Ok):
        return None
    header = verified.value.segments[-1].header
    thread = Thread(header.thread_id, header.branch_id, env.store)
    return Plan(
        "continued",
        turns=0,
        thread=thread,
        since=verified.value.fold.seq,
        base_cost_nanos=await cost_bound(thread),
    )


async def plan_prefix(case: CaseDir, log: CaseLog, target: EvalAgent, env: Env) -> Plan:
    texts = tuple(t for t in prefix_texts(log) if t is not None)
    if not texts:
        return Plan("none")
    same = dry_pin(target.definition).started.get("config_hash") == _recorded_hash(log)
    imported = await _import_case(case, env) if same else None
    if imported is None:
        return Plan("redriven", redrive=texts, turns=len(texts))
    return Plan(
        "continued",
        turns=len(texts),
        thread=imported.thread,
        since=imported.since,
        base_cost_nanos=imported.base_cost_nanos,
    )

File: typescript/packages/core/src/evals/simulate-prefix.ts
import type { Agent } from "../agent/agent";
import { dryPin } from "../agent/dry-pin";
import type { ThreadRef } from "../agent/result";
import { openStore } from "../agent/sqlite";
import type { CaseDir } from "./case-dir";
import type { CaseLog } from "./checks";
import { costBound, handleFor } from "./ledger";
import type { Env } from "./live-env";
import type { Simulation } from "./schema";

// The prefix of a simulated case (spec lane 32, B.1): the real turns before the saved one.
// Continue them when the agent's config_hash is unchanged -- the case log is imported and the
// conversation goes on, so the prefix costs no model call -- and re-drive their inputs when it
// changed, so a changed agent is graded on the whole path.

export type Plan = {
  readonly mode: Simulation["prefix"];
  /** Recorded inputs to re-send before the opener; empty unless the prefix is re-driven. */
  readonly redrive: readonly string[];
  readonly turns: number;
  /** The imported thread to continue; absent when the agent starts a new one. */
  readonly thread: ThreadRef | undefined;
  /** Events at or below this seq are the imported prefix, not this conversation's. */
  readonly since: number;
  readonly baseCostNanos: number;
};

/** The recorded user inputs before the saved turn, in log order. */
export const prefixTexts = (log: CaseLog): readonly (string | undefined)[] =>
  log.events.flatMap((e) => (e.type === "user_input" ? [e.data.text] : []));

const recordedHash = (log: CaseLog): string | undefined => {
  const started = log.events.find((e) => e.type === "thread_started");
  return started?.type === "thread_started"
    ? started.data.config_hash
    : undefined;
};

const NEW_THREAD = { thread: undefined, since: 0, baseCostNanos: 0 } as const;

/** Imports the case log into the eval store, so the conversation continues the real thread. */
async function importCase(
  c: CaseDir,
  env: Env,
): Promise<Pick<Plan, "thread" | "since" | "baseCostNanos">> {
  const store = await openStore(env.store);
  for (const artifact of c.artifacts) await store.artifacts.put(artifact);
  const imported = await store.log.importLog(c.log);
  const header = imported.ok
    ? imported.value.segments.at(-1)?.header
    : undefined;
  if (!imported.ok || header === undefined) return NEW_THREAD;
  const thread = {
    id: header.thread_id,
    branch: header.branch_id,
    store: env.store,
  };
  return {
    thread,
    since: imported.value.fold.seq,
    baseCostNanos: await costBound(await handleFor(env, thread)),
  };
}

export async function planPrefix(
  c: CaseDir,
  log: CaseLog,
  target: Agent<never, unknown>,
  env: Env,
): Promise<Plan> {
  const texts = prefixTexts(log).flatMap((t) => (t === undefined ? [] : [t]));
  if (texts.length === 0)
    return { mode: "none", redrive: [], turns: 0, ...NEW_THREAD };
  const same = dryPin(target).started.config_hash === recordedHash(log);
  const imported = same ? await importCase(c, env) : NEW_THREAD;
  return imported.thread === undefined
    ? { mode: "redriven", redrive: texts, turns: texts.length, ...NEW_THREAD }
    : { mode: "continued", redrive: [], turns: texts.length, ...imported };
}
